devices/flash-zynq7000/flashdrv.c

```diff
@@ -0,0 +1,461 @@
+	u8 *buff;
+	u32 buffPos;
+	u32 regID;
+	u32 sectID;
+
+	flash_info_t info;
+} fdrv_common;
+
+
+/* Auxiliary flash commands */
+
+static int flashdrv_cfiRead(flash_cfi_t *cfi)
+{
+	ssize_t res;
+	size_t cmdSz, dataSz;
+	/* RDID instruction is a generic for each flash memory */
+	const flash_cmd_t cmds = { .opCode = FLASH_CMD_RDID, .size = 1, .dummyCyc = 24 };
+
+	/* Cmd size is rounded to 4 bytes, it includes dummy cycles [b]. */
+	cmdSz = (cmds.size + cmds.dummyCyc / 8 + 0x3) & ~0x3;
+	/* Size of the data which is received during cmd transfer */
+	dataSz = cmds.dummyCyc / 8;
+
+	hal_memset(fdrv_common.cmdTx, 0, cmdSz);
+	fdrv_common.cmdTx[0] = cmds.opCode;
+
+	qspi_start();
+	if ((res = qspi_polledTransfer(fdrv_common.cmdTx, fdrv_common.cmdRx, cmdSz, TIMEOUT_CMD_MS)) < 0)
+		return res;
+
+	/* Copy data received after dummy bytes */
+	hal_memcpy((u8 *)cfi, (const void *)(fdrv_common.cmdRx + cmds.size), dataSz);
+
+	if ((res = qspi_polledTransfer(NULL, (u8 *)cfi + dataSz, sizeof(flash_cfi_t) - dataSz, 5 * TIMEOUT_CMD_MS)) < 0) {
+		qspi_stop();
+		return res;
+	}
+	qspi_stop();
+
+	return EOK;
+}
+
+
+static int flashdrv_statusRegGet(unsigned int cmdID, unsigned int *val)
+{
+	ssize_t res;
+	size_t cmdSz;
+	const flash_cmd_t *cmd;
+
+	if (cmdID != flash_cmd_rdsr1)
+		return -EINVAL;
+
+	*val = 0;
+	cmd = &fdrv_common.info.cmds[cmdID];
+
+	cmdSz = cmd->size + cmd->dummyCyc / 8;
+	hal_memset(fdrv_common.cmdTx, 0, cmdSz);
+	fdrv_common.cmdTx[0] = cmd->opCode;
+
+	qspi_start();
+	if ((res = qspi_polledTransfer(fdrv_common.cmdTx, (u8 *)val, cmdSz, TIMEOUT_CMD_MS)) < 0)
+		return res;
+	qspi_stop();
+
+	return EOK;
+}
+
+
+static int flashdrv_welSet(unsigned int cmdID)
+{
+	ssize_t res;
+	const flash_cmd_t *cmd;
+
+	if (cmdID != flash_cmd_wrdi && cmdID != flash_cmd_wren)
+		return -EINVAL;
+
+	cmd = &fdrv_common.info.cmds[cmdID];
+
+	hal_memset(fdrv_common.cmdTx, 0, cmd->size);
+	fdrv_common.cmdTx[0] = cmd->opCode;
+
+	qspi_start();
+	if ((res = qspi_polledTransfer(fdrv_common.cmdTx, fdrv_common.cmdRx, cmd->size, TIMEOUT_CMD_MS)) < 0)
+		return res;
+	qspi_stop();
+
+	return EOK;
+}
+
+
+static int flashdrv_regionFind(addr_t offs)
+{
+	int i;
+	size_t sz;
+	addr_t start = 0;
+	const flash_cfi_t *cfi = &fdrv_common.info.cfi;
+
+	for (i = 0; i < cfi->regsCount; ++i) {
+		sz = CFI_SIZE_REGION(cfi->regs[i].size, cfi->regs[i].count);
+		if (offs >= start && offs < sz)
+			return i;
+
+		start = sz;
+	}
+
+	return -EINVAL;
+}
+
+
+static int flashdrv_sectorErase(addr_t offs)
+{
+	int i;
+	ssize_t res;
+	time_t start, timeout;
+	unsigned int st;
+	const flash_cmd_t *cmd;
+	const flash_cfi_t *cfi = &fdrv_common.info.cfi;
+
+	if (offs > CFI_SIZE_FLASH(cfi->chipSize))
+		return -EINVAL;
+
+	if ((i = flashdrv_regionFind(offs)) < 0)
+		return i;
+
+	switch (CFI_SIZE_SECTION(cfi->regs[i].size)) {
+		case 0x1000:
+			cmd = &fdrv_common.info.cmds[flash_cmd_p4e];
+			break;
+		case 0x10000:
+			cmd = &fdrv_common.info.cmds[flash_cmd_se];
+			break;
```

Comment:
This whole thing looks like it's our Spantion memory specific. It might stay, but maybe there is some way to make this more generic?